### DISCUSSION TITLE
Performance Issues with large applications?

### DISCUSSION THREAD

**Discussion by alexkuttig** [software-mansion/react-native-reanimated#6559] [Q&A]
Hi, 

First of all - thanks for the great framework. I really like it and appreciate your contribution to the React Native ecosystem. 

I have a question regarding Reanimated performance in large apps.

I created a simple animation, which changes the style of my page header, consisting of an image and a text, based on a scroll view offset. I use a sharedValue and an animatedScrollHandler for the animation. The animation does not trigger anything else like state updates or other animations. 

This works fine when using in a single screen. But when using in our real world app, it starts to drop frames (UI and JS thread). Basically, the more screens are mounted, the slower the animation gets. It happens in release and dev mode and mainly on android. When I create the same animation with React Native's built in Animated API, this problem doesn't occur. It works stable at 60fps.

We are using the new architecture in our app and all examples here.

So my question is: Is there anything which can slow down reanimated when having a large apps (large Component Tree, many other animations, even if not running)? Is there anything we have to avoid, to prevent Reanimated from slowing down? 

We use React Navigation in our app and I already tried to use enableFreeze().

I would be super thankful for an answer and any tips/hints what could cause this performance problems with Reanimated.

**Comment by alexkuttig** [comment]
Hi again, 

I tested the same setup with React Native 0.76 and Reanimated 3.16.1. It works smooth on the old architecture, but not on the new architecture. React Native's built in Animated API works stable at 60fps on old and new architecture. 

Best, 
Alex

**Reply by mimo-10** [reply]
I have the same problem, reanimated team keeps ignoring this. There is also an issue created for that, they just close it like nothing is going on.

**Reply by alexkuttig** [reply]
Hi @mimo-10, we currently switched back to the old architecture, but prepare for switching to the new one. Since Reanimated still does not work for this use case, we switched to the Animated API for this kind of animation. Reanimated does work fine for all other Animations in our application, it's just the ones triggered by scroll. Maybe this can be a solution for you, too. 
We'll investigate Reanimated for this purpose from time to time, so I'll keep you posted here about our results.

**Reply by mimo-10** [reply]
Thank u so much @alexkuttig , appreciate it.

**Reply by Ayan4** [reply]
We are dealing with the same issue ever since we upgraded to react-native v0.76.6 and enabled the new architecture (reanimated v3.16.7). In the new arch, reanimated has made the app completely unusable because of frame drops in UI and JS thread. Unfortunately switching to react native's Animated API was not possible for us since scrollHandler was not the only animations affected.

**Reply by roo12312** [reply]
Facing the same issue yet reanimated team doesn't care

**Reply by mimo-10** [reply]
Yes, I said multiple times ,fixing this is way more important than shipping new features.

**Reply by roo12312** [reply]
Idk why reanimated team keeps ignoring this... Even after being reported multiple times... Probably their focus is just ios

**Reply by josegiufrida** [reply]
For us, the only thing preventing us from updating to the new architecture is the lag in Reanimated's animations.

Animations from our own or third-party components, especially in React-Native-Bottom-Sheet (changing the backdrop didn't work for us).

**Reply by tomekzaw** [reply]
@josegiufrida What do you mean by "lag in Reanimated's animations"? Can you share a video recording so we can investigate it?

**Reply by ConorLindsay94** [reply]
We are seeing the same issue anywhere we use `useAnimatedScrollHandler`. I thought this was a known issue that was being worked on? There are plenty of issues raised about similar problems on New Arch.

**Reply by tomekzaw** [reply]
@ConorLindsay94 What do you mean by "same issue"? Can you share a video recording so we can investigate it? Otherwise it's just complaining which is not helpful at all.

**Reply by alexkuttig** [reply]
Hi @tomekzaw, 
thanks for replying here. I'm the original author of this post and I can provide you everything you need. I don't want to "just complain". You're doing an awesome job and I really appreciate your open source work. 

I didn't find the videos from last time, but I'll create new ones today or tomorrow. Do you need anything else? I'm also available for a call if that would help.

**Reply by ConorLindsay94** [reply]
I mean we cannot turn New Arch on because a lot of our scroll animations rely on `useAnimatedScrollHandler`, and we see scroll issues elsewhere with Flash List etc. This is the only library we use which doesn't work well with New Arch, and there are plenty of other people saying the same thing.

https://github.com/software-mansion/react-native-reanimated/issues/7460
https://github.com/software-mansion/react-native-reanimated/issues/7108
https://github.com/software-mansion/react-native-reanimated/issues/6992
https://github.com/software-mansion/react-native-reanimated/issues/6708
https://github.com/software-mansion/react-native-reanimated/issues/7480

I don't want this to come across as complaining, and I appreciate you being so active on here @tomekzaw - just wanting some clarity on if there is any focus on performance issues people are seeing on New Arch, or if it is even in scope to be looked at at the minute?

In terms of a repro, I am working on other things at the minute because our app is stable on Old Arch. However I can get this for you as soon as I can.

**Reply by genesiscz** [reply]
@ConorLindsay94 As you mentioned my issue in #7460, team already mentioned what is happening and that the RN is kinda at fault there. I can also say that on iOS this started getting much worse too, as seen on this xCode regression graph:
<img width="832" alt="image" src="https://github.com/user-attachments/assets/cbbe39f7-5fa7-4218-aa81-7da3e7eb7632" />

**Reply by tomekzaw** [reply]
@alexkuttig No worries, thanks. There's several scroll-related issues happening at once. We need to see the actual behavior (and the exact versions of React Native and Reanimated, which you have already provided) so we can tell you which one it is and what you can do to fix it. Or, alternatively, you can wait 

@ConorLindsay94 Again, the term "scroll issues" isn't really helpful since this can mean various bugs. Also, I suspect it's not related to performance itself. By "performance", we mean the ability of Reanimated to animate multiple views at once and the actual number of FPS. It is likely that if you're facing "scroll issues", the app might still run on 60/120 fps but there can be some delay which causes the views to appear "laggy". Turns out, these two (namely "scroll issues" and "performance") are the current blockers of stable v4 and together with @bartlomiejbloniarz we're actively working on getting it fixed but we still have some other commitments like hosting App.js Conf 2025 last week. Please expect some updates on "scrolling issues" in the following days, but just so you know, it's likely that it's root cause is in react-native itself rather than Reanimated which makes things more tricky for all of us. As for FlashList, have you tried the new v2 version (`@shopify/flash-list@2.0.0-rc.2`)?

**Reply by mimo-10** [reply]
Please check #6710 for a video that show that, please check it : )

**Reply by mimo-10** [reply]
Hello any updates !? , at least tell us if u guys are working on this, this very very critical; (

**Reply by mimo-10** [reply]
btw, based oon you definition it is a performance issue, when i scroll especialy when i'm inside a page outside the man stack, fps is around 3 to 5 fps when an animation is triggered, with scrolling, thank u

**Reply by tomekzaw** [reply]
Hi all, if you're experiencing performance regressions of animations after enabling the New Architecture, please make sure to check out this page we recently added in Reanimated docs:
* [**ℹ️ Reanimated docs: Performance regressions on the New Architecture**](https://docs.swmansion.com/react-native-reanimated/docs/guides/performance#performance-regressions-on-the-new-architecture)

**Reply by ConorLindsay94** [reply]
Thank you for providing this @tomekzaw

**Comment by YueLiXing** [comment]
Can we negotiate with the RN team and officially release version 4.2.0 to solve the problem of using multiple flags

**Reply by tomekzaw** [reply]
We've already asked the React core team at Meta about enabling the `preventShadowTreeCommitExhaustion` feature flag by default but got no response so far.
